// iron_mq/lib/api_client.js
// Generated by CoffeeScript 1.6.2
(function() {
  var APIClient, ironCore, version, _,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  require('pkginfo')(module);

  version = this.version;

  _ = require('underscore');

  ironCore = require('iron_core');

  APIClient = (function(_super) {
    __extends(APIClient, _super);

    APIClient.prototype.AWS_US_EAST_HOST = 'mq-aws-us-east-1.iron.io';

    APIClient.prototype.RACKSPACE_HOST = 'mq-rackspace-dfw.iron.io';

    function APIClient(options) {
      var defaultOptions;

      defaultOptions = {
        scheme: 'https',
        host: this.AWS_US_EAST_HOST,
        port: 443,
        api_version: 1,
        user_agent: version,
        queue_name: 'default'
      };
      APIClient.__super__.constructor.call(this, 'iron', 'mq', options, defaultOptions, ['project_id', 'token', 'api_version', 'queue_name']);
    }

    APIClient.prototype.version = function() {
      return "iron_mq_node-" + version + " (" + (APIClient.__super__.version.call(this)) + ")";
    };

    APIClient.prototype.url = function() {
      return APIClient.__super__.url.call(this) + this.options.api_version.toString() + ("/projects/" + this.options.project_id + "/queues");
    };

    APIClient.prototype.headers = function() {
      return _.extend({}, APIClient.__super__.headers.call(this), {
        'Authorization': "OAuth " + this.options.token
      });
    };

    APIClient.prototype.queuesList = function(options, cb) {
      var parseResponseBind;

      parseResponseBind = _.bind(this.parseResponse, this);
      return this.get("", options, function(error, response, body) {
        return parseResponseBind(error, response, body, cb);
      });
    };

    APIClient.prototype.queuesGet = function(queue_name, cb) {
      var parseResponseBind;

      parseResponseBind = _.bind(this.parseResponse, this);
      return this.get("/" + queue_name, {}, function(error, response, body) {
        return parseResponseBind(error, response, body, cb);
      });
    };

    APIClient.prototype.queuesClear = function(queue_name, cb) {
      var parseResponseBind;

      parseResponseBind = _.bind(this.parseResponse, this);
      return this.post("/" + queue_name + "/clear", {}, function(error, response, body) {
        return parseResponseBind(error, response, body, cb);
      });
    };

    APIClient.prototype.queuesUpdate = function(queue_name, options, cb) {
      var parseResponseBind;

      parseResponseBind = _.bind(this.parseResponse, this);
      return this.post("/" + queue_name, options, function(error, response, body) {
        return parseResponseBind(error, response, body, cb);
      });
    };

    APIClient.prototype.queuesAddSubscribers = function(queue_name, subscribers, cb) {
      var parseResponseBind;

      parseResponseBind = _.bind(this.parseResponse, this);
      return this.post("/" + queue_name + "/subscribers", subscribers, function(error, response, body) {
        return parseResponseBind(error, response, body, cb);
      });
    };

    APIClient.prototype.queuesRemoveSubscribers = function(queue_name, subscribers, cb) {
      var parseResponseBind;

      parseResponseBind = _.bind(this.parseResponse, this);
      return this["delete"]("/" + queue_name + "/subscribers", subscribers, function(error, response, body) {
        return parseResponseBind(error, response, body, cb);
      });
    };

    APIClient.prototype.queuesDelete = function(queue_name, cb) {
      var parseResponseBind;

      parseResponseBind = _.bind(this.parseResponse, this);
      return this["delete"]("/" + queue_name, {}, function(error, response, body) {
        return parseResponseBind(error, response, body, cb);
      });
    };

    APIClient.prototype.messagesPost = function(queue_name, messages, cb) {
      var parseResponseBind;

      parseResponseBind = _.bind(this.parseResponse, this);
      return this.post("/" + queue_name + "/messages", {
        messages: messages
      }, function(error, response, body) {
        return parseResponseBind(error, response, body, cb);
      });
    };

    APIClient.prototype.messagesGet = function(queue_name, options, cb) {
      var parseResponseBind;

      parseResponseBind = _.bind(this.parseResponse, this);
      return this.get("/" + queue_name + "/messages", options, function(error, response, body) {
        return parseResponseBind(error, response, body, cb);
      });
    };

    APIClient.prototype.messagesDelete = function(queue_name, message_id, cb) {
      var parseResponseBind;

      parseResponseBind = _.bind(this.parseResponse, this);
      return this["delete"]("/" + queue_name + "/messages/" + message_id, {}, function(error, response, body) {
        return parseResponseBind(error, response, body, cb);
      });
    };

    APIClient.prototype.messagesPeek = function(queue_name, options, cb) {
      var parseResponseBind;

      parseResponseBind = _.bind(this.parseResponse, this);
      return this.get("/" + queue_name + "/messages/peek", options, function(error, response, body) {
        return parseResponseBind(error, response, body, cb);
      });
    };

    APIClient.prototype.messagesTouch = function(queue_name, message_id, cb) {
      var parseResponseBind;

      parseResponseBind = _.bind(this.parseResponse, this);
      return this.post("/" + queue_name + "/messages/" + message_id + "/touch", {}, function(error, response, body) {
        return parseResponseBind(error, response, body, cb);
      });
    };

    APIClient.prototype.messagesRelease = function(queue_name, message_id, options, cb) {
      var parseResponseBind;

      parseResponseBind = _.bind(this.parseResponse, this);
      return this.post("/" + queue_name + "/messages/" + message_id + "/release", options, function(error, response, body) {
        return parseResponseBind(error, response, body, cb);
      });
    };

    APIClient.prototype.messagesPushStatuses = function(queue_name, message_id, cb) {
      var parseResponseBind;

      parseResponseBind = _.bind(this.parseResponse, this);
      return this.get("/" + queue_name + "/messages/" + message_id + "/subscribers", {}, function(error, response, body) {
        return parseResponseBind(error, response, body, cb);
      });
    };

    APIClient.prototype.messagesPushStatusDelete = function(queue_name, message_id, subscriber_id, cb) {
      var parseResponseBind;

      parseResponseBind = _.bind(this.parseResponse, this);
      return this["delete"]("/" + queue_name + "/messages/" + message_id + "/subscribers/" + subscriber_id, {}, function(error, response, body) {
        return parseResponseBind(error, response, body, cb);
      });
    };

    return APIClient;

  })(ironCore.Client);

  module.exports.APIClient = APIClient;

}).call(this);
